### PR TITLE
Wat 537

### PR DESCRIPTION
**Pull request by ledamint**

### CODE REVIEW

File: packages/recorder-backend/src/path-composer.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    private targetAttribute: string;
+
+    constructor(attribute: string) {
+        this.targetAttribute = attribute;
+    }
+
+    getPath(elementsSummary: ElementSummary[]) {
+        return elementsSummary
+            .map((elemensSummary) => elemensSummary.attributes[this.targetAttribute])
+            .filter(attribute => attribute !== undefined)
```
Comment (johnthecat) [inline packages/recorder-backend/src/path-composer.ts]: different style of function declaration

File: packages/recorder-backend/src/recorder-server.ts [hunk 2]
```diff
@@ -33,6 +34,7 @@ export class RecorderServer implements IRecorderServer {
         private httpPort: number = DEFAULT_RECORDER_HTTP_PORT,
         private wsPort: number = DEFAULT_RECORDER_WS_PORT,
         private transportInstance: ITransport = transport,
+        private testManager: TestManager = new TestManager('SW'),
```
Comment (johnthecat) [inline packages/recorder-backend/src/recorder-server.ts]: Looks like hardcode, that we don't want to have
Discuss it with @flops

File: packages/recorder-backend/src/test-manager.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import { commandsByRecordingEventTypes, RecorderEvents } from '@testring/types';
+import { RECORDER_ELEMENT_IDENTIFIER } from '@testring/constants';
+
+export class TestManager extends EventEmitter {
+    constructor(
+        private manager: string,
+        private testWriter = new TestWriter(path.resolve(__dirname, 'test.js')),
+        private pathComposer = new PathComposer(RECORDER_ELEMENT_IDENTIFIER),
+    ) {
+        super();
```
Comment (johnthecat) [inline packages/recorder-backend/src/test-manager.ts]: unnecessary call of super

File: packages/recorder-extension/src/extension/elements-summary.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        return event.composedPath();
+    }
+
+    return composePath(event.target);
+};
+
+export const getElementsSummary = (elements: Array<Element>, withChildren: boolean = true): ElementSummary[] => {
+    return elements
+        .filter(({ tagName, attributes }) => tagName && attributes)
+        // @ts-ignore
```
[the remark this answers is not in the document]
Comment (ledamint) [reply packages/recorder-extension/src/extension/elements-summary.ts]: fixed

File: packages/recorder-backend/src/test-manager.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    private getCommand({ eventType, path, value, text }) {
+        return commandsByRecordingEventTypes[eventType]({ manager: this.manager, path, value, text });
+    }
+
+    // FIXME add type
```
[the remark this answers is not in the document]
Comment (ledamint) [reply packages/recorder-backend/src/test-manager.ts]: fixed

File: packages/recorder-backend/src/recorder-server.ts (hunk 2 above)
Comment (ledamint) [reply packages/recorder-backend/src/recorder-server.ts]: Yes, it was discussed, it's temporary

File: packages/recorder-backend/src/test-manager.ts (hunk 3 above)
Comment (ledamint) [reply packages/recorder-backend/src/test-manager.ts]: It extends by EventEmitter

File: packages/recorder-backend/src/path-composer.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { ElementSummary } from '@testring/types';
+
+export class PathComposer {
+    private targetAttribute: string;
+
+    constructor(attribute: string) {
```
[the remark this answers is not in the document]
Comment (ledamint) [reply packages/recorder-backend/src/path-composer.ts]: fixed

File: packages/recorder-backend/src/recorder-server.ts (hunk 2 above)
Comment (flops) [reply packages/recorder-backend/src/recorder-server.ts]: It's ok for me as mock. The bach-end is my part mostly.

**Comment by flops** [comment]
Taken here https://github.com/ringcentral/testring/pull/61

**Pull request closed without merging**